File: src/Poker/hand/FourOfAKind.java
package Poker.hand;

import Poker.game.DeckConstants.Value;

public class FourOfAKind extends Hand
{
   private Value fourKindValue;
   private Value kicker;
   
   public FourOfAKind(Value values[])
   {
      super(Rank.FOUR_KIND);
      fourKindValue = values[0];
      kicker = values[1];
   }
   
   public Value getFourKindValue()
   {
      return fourKindValue;
   }
   
   public Value getKicker()
   {
      return kicker;
   }
   
   public String toString()
   {
      return "4 of a kind - " + fourKindValue.toString() + "s";
   }
   
   public int compare(Hand handToCompare)
   {
      int iCompare = compareRank(handToCompare);
      if (iCompare != 0)
      {
         return iCompare;
      }
      
      return compareFourKinds((FourOfAKind) handToCompare);
   }
   
   private int compareFourKinds(FourOfAKind foak)
   {
      Value value = foak.getFourKindValue();
      if (fourKindValue != value)
      {
         return fourKindValue.getIndex() - value.getIndex();
      }
      
      return kicker.getIndex() - foak.getKicker().getIndex();
   }
}

File: server/hand/fourOfAKind.js
let Hand = require('./hand');

let FourOfAKind = (values) =>
{
	const FOUR_KIND = 7;
	
	let rank = Hand(FOUR_KIND);
	
	let fourKindValue = values[0];
	let kicker = values[1];
	
	let self =
	{
		getRank: () => rank,
		compare: (hand) =>
		{
			let rankCompare = rank.compare(hand.getRank());
			if (rankCompare)
			{
				return rankCompare;
			}
			
			return self.compareSameRank(hand);
		},
		compareSameRank: (fourOfAKind) =>
		{
			if (fourKindValue !== fourOfAKind.getFourKindValue())
			{
				return fourKindValue - fourOfAKind.getourKindValue();
			}
			
			return kicker - fourOfAKind.getKicker();
		},
		getFourKindValue: () => fourKindValue,
		getKicker: () => kicker,
		toString: () => 'Four of a kind (' + fourKindValue + "'s, " + kicker + ' kicker)'
	}
	
	return self;
}

module.exports = FourOfAKind;
